firedrake/formmanipulation.py

```diff
@@ -127,6 +128,59 @@ def argument(self, o):
                          for j in numpy.ndindex(V_is[i].value_shape)]
         return self._arg_cache.setdefault(o, as_vector(args))
 
+    def cofunction(self, o):
+        from firedrake import Cofunction
+        from ufl.classes import ZeroBaseForm
+
+        V = o.function_space()
+
+        # Not on a mixed space, just return ourselves.
+        if len(V) == 1:
+            return o
+
+        # We only need the test space for Cofunction
+        indices = self.blocks[0]
+        V_is = V.subfunctions
+
+        try:
+            indices = tuple(indices)
+            nidx = len(indices)
+        except TypeError:  # Only one index
+            indices = (indices, )
+            nidx = 1
+
+        # the cofunction should only be returned on the
+        # diagonal elements, so if we are off-diagonal
+        # on a two-form then we return a zero form,
+        # analogously to the off components of arguments.
+        if len(self.blocks) == 2:
+            itest, itrial = self.blocks[0], self.blocks[1]
+            on_diag = (itest == itrial)
```

Comment:
Is this handling `assemble(bilinear_form, diagonal=True)`? Is this covered by the tests?

Reply:
Do you mean 1) is this handling the case when `assemble(..., diagonal=True)`, or 2) will this work for `assemble(..., diagonal=True)`.

If 1) then no, this is handling the case where a diagonal block is being requested e.g. calling split with say `argument_indices=(1,1)`. The full form doesn't have to be diagonal.

If 2) then I don't know, I haven't tested it but I don't see why it wouldn't, this is just a symbolic step whereas it looks like the `diagonal` kwarg for `assemble` is an optimisation for the numerical assembly.

Reply:
@pbrubeck can I resolve this thread?